build_tools/linux_python_dist_split.py

```diff
@@ -155,31 +158,198 @@
+
+
+def get_os_arch() -> str:
+    """Gets the `os_arch` placeholder for the current system.
+
+    Matches the equivalent in the package _dist_info.py file.
+    """
+    return f"{platform.system().lower()}_{platform.machine()}"
+
 
 def run(args: argparse.Namespace):
     artifacts = ArtifactCatalog(args.artifact_dir)
     target_families = artifacts.all_target_families
+    if not target_families:
+        raise ValueError(f"No target artifacts found in {args.artifact_dir}")
+    default_target_family = sorted(target_families)[0]
+
+    # Prepare the _dist_info.py file by reading the template in the rocm-sdk
+    # package and adding lines to set distribution specific options.
+    dist_info_path = (
+        PYTHON_PACKAGING_DIR / "rocm-sdk" / "src" / "rocm_sdk" / "_dist_info.py"
+    )
+    dist_info_contents = dist_info_path.read_text()
+    dist_info_contents += f"__version__ = '{args.version}'\n"
+    dist_info_contents += f"PY_PACKAGE_SUFFIX_NONCE = '{args.version_suffix}'\n"
+    dist_info_contents += f"DEFAULT_TARGET_FAMILY = '{default_target_family}'\n"
+    for target_family in target_families:
+        dist_info_contents += f"AVAILABLE_TARGET_FAMILIES.append('{target_family}')\n"
+    log(f"_dist_info.py: {dist_info_contents}", vlog=1)
+
+    populate_built_artifacts(
+        args,
+        artifacts=artifacts,
+        target_families=target_families,
+        dist_info_contents=dist_info_contents,
+    )
+
+    if args.build_packages:
+        build_packages(args, args.dest_dir)
+
+
+def build_packages(args: argparse.Namespace, dest_dir: Path):
+    dist_dir = dest_dir / "dist"
+    for child_path in dest_dir.iterdir():
+        if not child_path.is_dir():
+            continue
+        if not (child_path / "pyproject.toml").exists():
+            continue
+        child_name = child_path.name
+
+        # Some of our packages build as sdists and some as wheels.
+        # Contrary to documented wisdom, we invoke setuptools directly. This is
+        # because the "build frontends" have an impossible compatibility matrix
+        # and opinions about how to pass arguments to the backends. So we skip
+        # the frontends for such a closed case as this.
+        build_args = [sys.executable, "-m", "build", "-v", "--outdir", str(dist_dir)]
+        build_args = [
+            sys.executable,
+            str(child_path / "setup.py"),
+        ]
+        if child_name in ["rocm-sdk"]:
+            build_args.append("sdist")
+        else:
+            build_args.append("bdist_wheel")
+            if not args.wheel_compression:
+                build_args.append("--compression")
+                build_args.append("stored")
+        build_args.extend(
+            [
+                "-v",
+                "--dist-dir",
+                str(dist_dir),
+            ]
+        )
+
+        log(f"::: Building python package {child_name}: {shlex.join(build_args)}")
+        subprocess.check_call(build_args, cwd=child_path)
```

Comment:
I might not use the script correctly but in my first attempt I passed a relative `--dest-dir` pointing to `packages` located in the root of the repo. This failed here with
> /home/mbrehler/TheRock/venv-2/bin/python: can't open file '/home/mbrehler/TheRock/packages/rocm-sdk/packages/rocm-sdk/setup.py': [Errno 2] No such file or directory

as the subprocess has set the current working directory to `child_path`. Passing an absolute path (or rather passing``--dest-dir `realpath packages` ``) solved this here but this is something others might trap into as well.

Reply:
Oh thanks. I should just absolute the path immediately to Nate it resilient.